sparsify/trainer.py

```diff
@@ -82,46 +82,77 @@ def __init__(
-            for sae in self.saes.values()
-        ]
-        # Dedup the learning rates we're using, sort them, round to 2 decimal places
-        lrs = [f"{lr:.2e}" for lr in sorted(set(pg["lr"] for pg in pgs))]
-        print(f"Learning rates: {lrs}" if len(lrs) > 1 else f"Learning rate: {lrs[0]}")
 
-        if cfg.optimizer == 'adam':
-            try:
-                from bitsandbytes.optim import Adam8bit as Adam
+        assert isinstance(dataset, Sized)
+        num_batches = len(dataset) // cfg.batch_size
 
-                print("Using 8-bit Adam from bitsandbytes")
-            except ImportError:
-                from torch.optim import Adam
+        match cfg.optimizer:
+            case "adam":
+                try:
+                    from bitsandbytes.optim import Adam8bit as Adam
 
-                print("bitsandbytes 8-bit Adam not available, using torch.optim.Adam")
-                print("Run `pip install bitsandbytes` for less memory usage.")
+                    print("Using 8-bit Adam from bitsandbytes")
+                except ImportError:
+                    from torch.optim import Adam
 
-            assert isinstance(dataset, Sized)
-            num_examples = len(dataset)
+                    print(
+                        "bitsandbytes 8-bit Adam not available, using torch.optim.Adam"
+                    )
+                    print("Run `pip install bitsandbytes` for less memory usage.")
 
-            self.optimizer = Adam(pgs)
-            self.lr_scheduler = get_linear_schedule_with_warmup(
-                self.optimizer, cfg.lr_warmup_steps, num_examples // cfg.batch_size
-            )
-        else:
-            self.optimizer = ScheduleFreeWrapper(SignSGD(pgs), momentum=0.95)
-            self.optimizer.train()
-            self.lr_scheduler = None
-            
+                pgs = [
+                    dict(
+                        params=sae.parameters(),
+                        lr=cfg.lr or 2e-4 / (sae.num_latents / (2**14)) ** 0.5,
+                    )
+                    for sae in self.saes.values()
+                ]
+                # For logging purposes
+                lrs = [f"{lr:.2e}" for lr in sorted(set(pg["lr"] for pg in pgs))]
+
+                adam = Adam(pgs)
+                self.optimizers = [adam]
+                self.lr_schedulers = [
+                    get_linear_schedule_with_warmup(
+                        adam, cfg.lr_warmup_steps, num_batches
+                    )
+                ]
+            case "muon":
+                params = {p for sae in self.saes.values() for p in sae.parameters()}
+                muon_params = {p for p in params if p.ndim >= 2}
+                lrs = [f"{cfg.lr or 2e-3:.2e}"]
+
+                self.optimizers = [
+                    # Muon LR is independent of the number of latents
+                    Muon(muon_params, lr=cfg.lr or 2e-3),
+                    torch.optim.Adam(params - muon_params, lr=cfg.lr or 2e-3),
```

Comment:
Do we want to try to use bits and bytes Adam here?

Reply:
I figure it doesn't matter since the biases are super tiny